src/ScheduleFunctions.cpp

```diff
@@ -79,6 +79,58 @@ bool contains_impure_call(const Expr &expr) {
     return is_not_pure.result;
 }
 
+// A mutator that performs a substitute operation only for a certain kind
```

Comment:
I don't understand how this works when IRNodeType is anything other than Provide. Doesn't it recursively substitute things inside the values?

Reply:
Oh it only calls substitute on the args, not the values. That's non-obvious.

Reply:
What happens if you predicate stores on:

`hist(im(r.x, r.y)) += 1`

when splitting one of the rvars?

I think this mutator might do the wrong thing in that scenario.

Reply:
I decided to just disable PredicateStores for reductions, because it's pretty confusing even if it works. (I already didn't enable PredicateLoads).

Reply:
For reductions, PredicateStores seems fine to me, even if it implies unpredicated self-loads. No stores = no side-effects.

Reply:
I was mostly using that example not to complain about reductions, but to point out that this mutator doesn't really handle nesting. E.g. if a call is found inside the LHS of a Provide. Passing it arbitrary IRNodeTypes is going to break whenever they can nest.

Another issue is if the things that need to be substituted out have already been lifted out of the index of a Call into a let by the CSE pass that happens inside Function::define to protect us from pathological Exprs. Or even worse:

```foo[let y = the_loop_var in bar[y] + y] = 5```

Yuck. OK, maybe disallowing reductions is wise for now.

I think the correct thing (with reductions forbidden) is that for predicate loads, you just substitute in the entire RHS of each provide, and for predicate stores, you substitute in the entire LHS of each provide.